### ISSUE TITLE
Startup fails (on kubernetes)

### ISSUE THREAD

**Issue by elm-** [joeferner/redis-commander#267]
I'm getting the following error when running it on Kubernetes. The config looks correct to me, anyone has an idea how to check this further?

```
Creating custom redis-commander config.
Configuration:
    {
    "sidebarWidth":250,
    "locked":false,
    "CLIHeight":50,
    "CLIOpen":false,
    "default_connections": [
        {
            "label":"redis",
            "host":"redis",
            "port":"6379",
            "password":"",
            "dbIndex":0
        }
        ]
    }
node ./bin/redis-commander --redis-port tcp://10.47.248.71:6379
No Save: false
listening on  0.0.0.0 : 8081
net.js:1032
      throw new errors.RangeError('ERR_SOCKET_BAD_PORT', port);
      ^

RangeError [ERR_SOCKET_BAD_PORT]: Port should be > 0 and < 65536. Received NaN.
    at lookupAndConnect (net.js:1032:13)
    at Socket.connect (net.js:1001:5)
    at Object.connect (net.js:124:35)
    at /redis-commander/node_modules/ioredis/lib/connectors/connector.js:44:20
    at process._tickCallback (internal/process/next_tick.js:176:11)
```

The config for it:

```
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: redis-commander
spec:
  replicas: 1
  template:
    metadata:
      labels:
        app: airflow
        tier: redis-commander
    spec:
      restartPolicy: Always
      containers:
      - name: redis
        image: rediscommander/redis-commander:latest
        ports:
        - name: redis-commander
          containerPort: 8081
        env:
        - name: REDIS_HOSTS
          value: redis:redis:6379
```

**Comment by stevenaldinger** [comment]
I looked into it the best I could and was not able to reproduce. **However, you gave us enough context that I can tell you what's going on and hopefully help you hack a solution**.

TLDR:

**Try adding this to your k8s spec (explicitly setting REDIS_PORT to empty string)**. If you only have a single redis instance running you can _maybe_ get away with REDIS_PORT=6379 but setting that explicitly will break other things I believe. I wrote the original entrypoint.sh and redis commander was not compatible with both configuration files and command line args (to my knowledge) so I just generated the config with env vars to give flexibility. Someone else has altered it since and there are a lot of caveats I don't think they considered.

```
        env:
        - name: REDIS_HOSTS
           value: redis:redis:6379
        - name: REDIS_PORT
           value: ""
```

1. This is the minimal equivalent to what you're doing which works fine:
  - `docker run -it -e REDIS_HOSTS=redis:redis:6379 rediscommander/redis-commander:latest`
2. I also ran the exact config you supplied and it worked fine.

**The key to your mystery is that your logs show `node ./bin/redis-commander --redis-port tcp://10.47.248.71:6379`.** [This is the only line in the entrypoint.sh script](https://github.com/joeferner/redis-commander/blame/master/docker/entrypoint.sh#L135) that is able to cause that behavior, given you clearly don't have overrides in your k8s spec (that would screw with the entrypoint).

My command (try running this):
> $:- kubectl exec -it "$(kubectl get po -l app=airflow -l tier=redis-commander -o jsonpath='{.items[0].metadata.name}')" -- sh -c "env | grep REDIS_"

My output:
> REDIS_HOSTS=redis:redis:6379
What I think your output is going to show:
> REDIS_HOSTS=redis:redis:6379
> REDIS_PORT=tcp://10.47.248.71:6379

**And to try the theory backwards to prove that _would_ case the behavior...**

My command:
> $: docker run -it -e REDIS_HOSTS=redis:redis:6379 -e REDIS_PORT=tcp://10.47.248.71:6379 rediscommander/redis-commander:latest
Output:
```
Creating custom redis-commander config.
Configuration:
    {
    "sidebarWidth":250,
    "locked":false,
    "CLIHeight":50,
    "CLIOpen":false,
    "default_connections": [
        {
            "label":"redis",
            "host":"redis",
            "port":"6379",
            "password":"",
            "dbIndex":0
        }
        ]
    }
node ./bin/redis-commander --redis-port tcp://10.47.248.71:6379
No Save: false
listening on  0.0.0.0 : 8081
setUpConnection Redis error RangeError [ERR_SOCKET_BAD_PORT]: Port should be > 0 and < 65536. Received NaN.
    at lookupAndConnect (net.js:1032:13)
    at Socket.connect (net.js:1001:5)
    at Object.connect (net.js:124:35)
    at /redis-commander/node_modules/ioredis/lib/connectors/connector.js:47:22
    at process._tickCallback (internal/process/next_tick.js:176:11)
```

**Comment by elm-** [comment]
Hi, sorry for the late reply, yes, that was the reason. Kubernetes sets by default environment variables for all defined ports and services, e.g. a service of name "redis" will generate in all pods in that namespace "REDIS_PORT" with the full IP and port as a variable. Can only be override with the pod definition. I specified REDIS_PORT as you suggested and it worked. Was just a clash in testing anyway as in production there will be a prefix, e.g. "prod-airflow-redis" on the service, but good to know.

Thanks for the reply and details.

**Comment by rusenask** [comment]
Hi, ran into the same problem, your solution of adding empty string for port worked. Maybe you could update official docs? What k8s version are you running? Mine is 1.10

**Comment by klmcwhirter** [comment]
I can confirm this. I had a different problem deploying into OpenShift but the resolution was the same.

I was setting the following env vars: REDIS_HOST, REDIS_PORT.
i was getting an error along the lines of "could not write to /redis-commander/.redis-commander" due to entrypoint.sh: writeDefaultConfigBeginning function.

The workaround was to delete those env vars and define REDiS_HOSTS=myproject:redis:6379

I know the issue is closed, but wanted to document this in case anyone else encounters this with OpenShift.

This was in Minishift 1.27 with oc 3.11 and K8S 1.11.
minishift v1.27.0+707887e

oc v3.11.0+0cbc58b
kubernetes v1.11.0+d4cacc0
features: Basic-Auth GSSAPI Kerberos SPNEGO

Server https://192.168.99.101:8443
kubernetes v1.11.0+d4cacc0

**Comment by sseide** [comment]
@kimcwhirter this is strange as setting `REDIS_HOSTS` var does write this config file too with the same code. The entry-point.sh calls function `parseRedisHosts` which - in turn - calls `writeDefaultConfigBeginning`...

We only use the REDIS_HOST/REDIS_PORT/... env vars without REDIS_HOSTS inside a kubernetes cluster without any problems...

But thanks for sharing this.

**Comment by klmcwhirter** [comment]
This is what I am seeing in the log for the OpenShift pod.

If seems the root cause is some missing permissions management in the
Dockerfile - but am not sure.


/redis-commander/docker/entrypoint.sh: line 124: can't create
/redis-commander/.redis-commander: Permission denied
Creating custom redis-commander config.
/redis-commander/docker/entrypoint.sh: line 124: can't create
/redis-commander/.redis-commander: Permission denied
/redis-commander/docker/entrypoint.sh: line 124: can't create
/redis-commander/.redis-commander: Permission denied
/redis-commander/docker/entrypoint.sh: line 131: can't create
/redis-commander/.redis-commander: Permission denied
/redis-commander/docker/entrypoint.sh: line 131: can't create
/redis-commander/.redis-commander: Permission denied
cat: can't open '/redis-commander/.redis-commander': No such file or
directory
Creating custom redis-commander config.
Configuration:
node ./bin/redis-commander --redis-port tcp://172.30.93.8:6379
--redis-password 5C4nL558mO2piFhO
{ Error: ENOENT: no such file or directory, open
'/redis-commander/.redis-commander'
errno: -2,
code: 'ENOENT',
syscall: 'open',
path: '/redis-commander/.redis-commander' }
No config found or was invalid.
Using default configuration.
No Save: false
setUpConnection Redis error RangeError: "port" option should be >= 0 and <
65536: NaN
at lookupAndConnect (net.js:1052:13)
at Socket.connect (net.js:1025:5)
at Object.connect (net.js:117:35)
at /redis-commander/node_modules/ioredis/lib/connectors/connector.js:47:22
at _combinedTickCallback (internal/process/next_tick.js:132:7)
at process._tickCallback (internal/process/next_tick.js:181:9)
Connection closed. Attempting to Reconnect...
listening on 0.0.0.0 : 8081
Problem saving config.
{ Error: EACCES: permission denied, open
'/redis-commander/.redis-commander'
errno: -13,
code: 'EACCES',
syscall: 'open',
path: '/redis-commander/.redis-commander' }

**Comment by sseide** [comment]
Regarding permissions in Dockerfile:

https://github.com/joeferner/redis-commander/blob/c6ad89a5c9af694ae3e91355a2e7b7c06821875b/Dockerfile#L24-L25

The HOME var is set before to the `/redis-commander` directory, therefore all files within and the directory itself is owned by the user running redis-commander (USER command at line 32) below. 

The docker image itself should be fine. Are there some special restrictions regarding OpenShift on writing to the image? Maybe something like the --read-only flag to docker run? (https://docs.docker.com/engine/reference/commandline/run/)

Current behaviour of redis-commander depends on writing the file if its not present (default build) or the `REDIS_HOSTS` env var is set.

Other possibility to work around may be injecting the .redis-commander config file from Openshift with all connection configured inside the `default_connections` List and not setting the ENV vars et all...

**Comment by sseide** [comment]
Addon: a sample config file can be found at `docker/redis-commader.json`

**Comment by klmcwhirter** [comment]
I had been using the latest version in Aug / Sep and it was working fine
using the REDIS_HOST / REDIS_PORT env vars in OpenShift.

It is not working now.

So something changed in the redis-commander code since then to break it in
OpenShift.

FYI I have tried it in both 3.9 and 3.11 with the same result.

At least with the REDIS_HOSTS env var it will start - the config is wrong
so it does not function - falls back to default config. But with the
individual env vars it will not even start.


Steps to reproduce:
pre) install MiniShift 1.27 -
https://github.com/minishift/minishift/releases

1) Create redis instance using defaults - I am using this template:
https://github.com/klmcwhirter/Huntwords.PuzzleBoard.Cache.Manager/blob/master/etc/redis/redis-persistent-template.json

2) Add to project: Deploy Image: docker.io/rediscomander/redis-commander
Set env var REDIS_HOSTS myproject:redis:6379

Hope that helps.

-k

**Comment by sseide** [comment]
Sorry - its an Openshift problem. Openshift ignores the USER command from the Dockerfile (or similiar). You either has to find out how this can be configured inside Openshift or create your own image.

Running the image through docker or plain kubernetes and attaching to the running image with "docker exec -it"  or similiar:
```~ $ id
uid=100(redis) gid=65533
~ $ pwd
/redis-commander
~ $ touch .test
~ $ ls -la
total 184
drwxr-sr-x    1 redis    65533           63 Nov 22 14:15 .
drwxr-xr-x    1 root     root            29 Nov 22 14:13 ..
-rw-------    1 redis    65533           33 Nov 22 14:15 .ash_history
drwx--S---    3 root     65533           25 Nov 22 12:33 .config
drwxr-sr-x    5 redis    65533           86 Nov 22 12:33 .npm
-rw-r--r--    1 redis    65533          290 Nov 22 14:13 .redis-commander
-rw-r--r--    1 redis    root           233 Jul 24 14:50 .redis_commander
-rw-r--r--    1 redis    65533            0 Nov 22 14:15 .test
-rw-r--r--    1 redis    root          1083 Jul 24 14:50 LICENSE
drwxr-xr-x    1 redis    root            32 Nov 13 13:56 bin
drwxr-xr-x    1 redis    root            95 Nov 22 12:32 docker
-rw-r--r--    1 redis    root           488 Jul 24 14:50 ecosystem.config.js
drwxr-xr-x    1 redis    root            49 Nov 13 13:56 lib
drwxr-sr-x  334 root     65533        12288 Nov 22 12:33 node_modules
-rw-r--r--    1 redis    root        147162 Nov 22 12:33 package-lock.json
-rw-r--r--    1 redis    root          1342 Nov 13 13:56 package.json
drwxr-xr-x    1 redis    root            33 Sep 11 14:00 web
~ $
```

Running the same image inside minishift and doing an "docker exec -it":
```/redis-commander $ id
uid=1000170000 gid=0(root) groups=1000170000
/redis-commander $ pwd
/redis-commander
/redis-commander $ touch .test
touch: .test: Permission denied
/redis-commander $ ls -la
total 176
drwxr-sr-x    1 redis    65533          211 Nov 21 15:01 .
drwxr-xr-x    1 root     root            17 Nov 22 14:10 ..
drwx--S---    3 root     65533           25 Nov 21 15:01 .config
drwxr-sr-x    5 redis    65533           86 Nov 21 15:01 .npm
-rw-r--r--    1 redis    root           233 Nov 21 15:00 .redis_commander
-rw-r--r--    1 redis    root          1083 Nov 21 15:00 LICENSE
drwxr-xr-x    1 redis    root            32 Nov 21 15:00 bin
drwxr-xr-x    1 redis    root            95 Nov 21 15:00 docker
-rw-r--r--    1 redis    root           488 Nov 21 15:00 ecosystem.config.js
drwxr-xr-x    1 redis    root            49 Nov 21 15:00 lib
drwxr-sr-x  334 root     65533        12288 Nov 21 15:01 node_modules
-rw-r--r--    1 redis    root        147162 Nov 21 15:01 package-lock.json
-rw-r--r--    1 redis    root          1342 Nov 21 15:00 package.json
drwxr-xr-x    1 redis    root            33 Nov 21 15:00 web
/redis-commander $
```
As you see inside minishift the user has another id (not 100 anymore) and therefor no rights to write to the folder.

It may have worked with images from before 19 june 2018. Up to this date the app inside the container was running as "root", which is from a security point of view a no-go. With that commit  it was changed inside Dockerfile to "USER redis". And (not using Openshift) this user has rights to create files inside its own home.

Another possibility may be to move the evaluation of REDIS_HOSTS var inside the node app without creating a config file first...

**Comment by sseide** [comment]
regarding Openshift problem - one may say so - some other not... :-)
At least it is mentioned under "OpenShift Container Platform-Specific Guidelines"

Afair the app does no user to id lookup, therefor the passwd specific parts may be omited.
Only root group for home directory and numeric USER id should be needed (as i read it).

For numeric user id it should be testet if an env var is possible too (create user, get id of newly created user and store it inside an env var and use that for the USER statement to not fail if (for whatever reason) the id of newly create users change during build process.

When you change the Dockerfile - maybe you can do some extra steps to increase security:
only HOME dir itself (/redis-commander) needs to be writable for the user to create config file, node_modules / web / lib / ... folder and so on can belong to root to not allow some changes during runtime from inside the app.

Any PR fixing problems is welcome i think...
